src/Nethermind/Nethermind.Blockchain/Producers/BlockProducer.cs

```diff
@@ -0,0 +1,307 @@
+            ILogManager? logManager,
+            IDifficultyCalculator? difficultyCalculator)
+        {
+            _txSource = txSource ?? throw new ArgumentNullException(nameof(txSource));
+            Processor = processor ?? throw new ArgumentNullException(nameof(processor));
+            Sealer = sealer ?? throw new ArgumentNullException(nameof(sealer));
+            BlockTree = blockTree ?? throw new ArgumentNullException(nameof(blockTree));
+            StateProvider = stateProvider ?? throw new ArgumentNullException(nameof(stateProvider));
+            _gasLimitCalculator = gasLimitCalculator ?? throw new ArgumentNullException(nameof(gasLimitCalculator));
+            Timestamper = timestamper ?? throw new ArgumentNullException(nameof(timestamper));
+            _specProvider = specProvider ?? throw new ArgumentNullException(nameof(specProvider));
+            _trigger = trigger ?? throw new ArgumentNullException(nameof(trigger));
+            _difficultyCalculator = difficultyCalculator ?? throw new ArgumentNullException(nameof(difficultyCalculator));
+            Logger = logManager?.GetClassLogger() ?? throw new ArgumentNullException(nameof(logManager));
+        }
+
+        private void OnTriggerBlockProduction(object? sender, BlockProductionEventArgs e)
+        {
+            e.BlockProductionTask = TryProduceAndAnnounceNewBlock(e.CancellationToken, e.ParentHeader);
+        }
+
+        public virtual void Start()
+        {
+            _producerCancellationToken = new CancellationTokenSource();
+            _isRunning = true;
+            _trigger.TriggerBlockProduction += OnTriggerBlockProduction;
+            _lastProducedBlockDateTime = DateTime.UtcNow;
+        }
+
+        public virtual Task StopAsync()
+        {
+            _producerCancellationToken?.Cancel();
+            _isRunning = false;
+            _trigger.TriggerBlockProduction -= OnTriggerBlockProduction;
+            _producerCancellationToken?.Dispose();
+            return Task.CompletedTask;
+        }
+
+        protected virtual bool IsRunning() => _isRunning;
+
+        public bool IsProducingBlocks(ulong? maxProducingInterval)
+        {
+            if (Logger.IsTrace) Logger.Trace($"Checking IsProducingBlocks: maxProducingInterval {maxProducingInterval}, _lastProducedBlock {_lastProducedBlockDateTime}, IsRunning() {IsRunning()}");
+            return IsRunning() && (maxProducingInterval == null || _lastProducedBlockDateTime.AddSeconds(maxProducingInterval.Value) > DateTime.UtcNow);
+        }
+
+        private async Task<Block?> TryProduceAndAnnounceNewBlock(CancellationToken token, BlockHeader? parentHeader = null)
+        {
+            using CancellationTokenSource tokenSource = CancellationTokenSource.CreateLinkedTokenSource(token, _producerCancellationToken!.Token);
+            token = tokenSource.Token;
+            
+            Block? block = null;
+            if (await _producingBlockLock.WaitOneAsync(1000, token))
```

Comment:
Can we add a comment to what happens when it is false? Maybe drbug diagnostic then?